src/relations/openfga.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Define the Temporal server openfga relation."""
+
+import json
+import logging
+
+import requests
+from charms.openfga_k8s.v0.openfga import OpenFGAStoreCreateEvent
+from ops import framework
+
+from log import log_event_handler
+
+logger = logging.getLogger(__name__)
+
+
+class OpenFGA(framework.Object):
+    """Client for openfga:temporal relations."""
+
+    def __init__(self, charm):
+        """Construct.
+
+        Args:
+            charm: The charm to attach the hooks to.
+        """
+        super().__init__(charm, "openfga")
+        self.charm = charm
+        # Register OpenFGA relation handlers.
+        charm.framework.observe(
+            charm.openfga.on.openfga_store_created,
+            self._on_openfga_store_created,
+        )
+        charm.framework.observe(
+            charm.on.create_authorization_model_action,
+            self._on_create_authorization_model_action,
+        )
+        charm.framework.observe(charm.on.openfga_relation_broken, self._on_openfga_relation_broken)
+
+    @log_event_handler(logger)
+    def _on_openfga_store_created(self, event: OpenFGAStoreCreateEvent):
+        """Handle OpenFGA relation created event.
+
+        Args:
+            event: The event triggered when the relation is created.
+        """
+        if not event.store_id:
+            logger.info(f"{event.relation.name} revoked, no store id")
+            return
+
+        token = event.token
+        if event.token_secret_id:
+            secret = self.charm.model.get_secret(id=event.token_secret_id)
+            secret_content = secret.get_content()
+            token = secret_content["token"]
+
+        if self.charm.unit.is_leader():
+            self.charm._state.openfga = {
+                "store_id": event.store_id,
+                "token": token,
+                "address": event.address,
+                "port": event.port,
+                "scheme": event.scheme,
+                "auth_model_id": None,
+            }
+
+        self.charm._update(event)
+
+    @log_event_handler(logger)
+    def _on_openfga_relation_broken(self, event) -> None:
+        """Handle broken relations with OpenFGA.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        if not self.charm._state.is_ready():
+            event.defer()
+            return
+
+        if self.charm.unit.is_leader():
+            self.charm._state.openfga = None
+            self.charm._update(event)
+
+    @log_event_handler(logger)
+    def _on_create_authorization_model_action(self, event):
+        """Handle OpenFGA create authorization model action.
```

Comment:
Not a fan of this function's length, but it's up to you to refactor it to more readable sections.

Reply:
I've refactored this into smaller, more readable functions.